libbeat/processors/add_kubernetes_metadata/matchers.go

```diff
@@ -173,3 +177,94 @@ func (f *FieldFormatMatcher) MetadataIndex(event common.MapStr) string {
 
 	return string(bytes)
 }
+
+// PidMatcher implements matcher that matches container id based on pid from cgroup file
+type PidMatcher struct {
+	matchRegex  string
+	pidCidCache *common.Cache
+}
+
+// NewPidMatcher initializes and returns a PidMatcher
+func NewPidMatcher(cfg common.Config) (Matcher, error) {
+	config := struct {
+		RegexField string `config:"matcher_regex"`
```

Comment:
Since k8s is following a principle of `/prefix/class/podID/cID` wouldn't string manipulation be more performant?
e.g.:
```golang
func getCID(cgroup string) (string, error) {
	if i := strings.LastIndex(cgroup, "/"); i >= 0 {

		return cgroup[i+1:], nil

	}
	return "", fmt.Errorf("Can't extract container ID, cgroup: %s", cgroup)
}
```

Reply:
Actually I do not know about performance, i'm sure regex is more flexible. 
For example this another PR used prefixes from conf instead to choose right line from cgroups returned by gosigar...

Reply:
would you prefer using gosigar and prfixes like other PR does?

Reply:
As far as I understand gosigar will get you the content of `/proc/$pid/cgroup`.

https://github.com/elastic/beats/blob/master/libbeat/processors/add_docker_metadata/add_docker_metadata.go#L259

you use prefixes for detecting if given process is using e.g. `/docker` or `/kubepods`

check my PR to `add_docker_metadata`: https://github.com/elastic/beats/pull/16926

Reply:
Yes, i understand how it works. gosigar gives map of cgroup strings, but i'm not able to assess performance difference over that current regex solution where i search first regex match over whole cgroup file content...

If you know or feel that iterating over that string map searching matching prefix and then splitting string is faster, i'm happy to change.

Just a thought, this kubernetes meta plugin should be used only with kubernetes, so i guess that docker specific setup is not relevant here...

All together I'd like to end up with good solution :) Sadly i'm completely new with go and also beats, just solving a problem for related project my work depends on. So all of your advice is more than appreciated. 

Thank you!!!

Reply:
only time I'm iterating over whole cgroup map is when there is no container ID otherwise only the first entry is checked. I'll read a little bit more about cgroups and containers because I don't know if there is a possibility that it can run in a container with only subset of cgroups created

Reply:
https://medium.com/@dgryski/speeding-up-regexp-matching-with-ragel-4727f1c16027

Here is something about regex performance....